File: extension/edr/extension/WatchedPaths.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.fleetdm.edr.securityextension", category: "WatchedPaths")

/// WatchedPathMatch is how a watched path covers files, and maps one-to-one onto the Endpoint Security target mute types the
/// file-tamper client applies it with.
enum WatchedPathMatch: String, Sendable {
    /// literal covers exactly the file at the path.
    case literal
    /// prefix covers every path that starts with it, at any depth. A prefix meant as a directory ends in "/", or it would also
    /// cover a sibling that merely shares the leading characters (`/etc/sudoers.d` would match `/etc/sudoers.d.bak`).
    case prefix
}

/// WatchedPath is one entry in the set of file paths the file-tamper client observes.
struct WatchedPath: Hashable, Sendable {
    let path: String
    let match: WatchedPathMatch
}

/// WatchedPathsUpdate is a decoded `watched_paths.update`: the server's version and epoch for the set, the entries this extension
/// applies, and how many entries it skipped.
struct WatchedPathsUpdate: Equatable {
    let version: Int64
    /// epoch is the set's server update time in Unix microseconds, 0 when the server sent none. It orders sets across a server
    /// database restore that regresses version, the same role policy_epoch plays for application control (#322).
    let epoch: Int64
    let paths: [WatchedPath]
    let skipped: Int

    /// supersedes reports whether this update should replace `current`, the last one accepted. Sets are ordered by epoch, then by
    /// version: the server forces each update time past the one in its database, so epoch orders every set that database issues,
    /// and after a restore sends version backwards the next set is ahead once the database clock is past any epoch the restore
    /// lost. Version breaks a tie, which is what orders sets from a server that sends no epoch. An update that is not ahead is a
    /// delayed or duplicate delivery, and applying it would put an older set back over a newer one. Commands can reach the agent
    /// out of order (a poll returns pending commands newest first, and the poll and the control stream can overlap), so this is
    /// the gate that keeps the newest set in force. Application control orders its snapshots the same way
    /// (ApplicationControlStore.apply).
    func supersedes(_ current: WatchedPathsUpdate?) -> Bool {
        guard let current else { return true }
        return PushOrder(epoch: epoch, version: version) > PushOrder(epoch: current.epoch, version: current.version)
    }
}

/// WatchedPaths is the file-tamper client's watched set (ADR-0008 step 4, issue #998): which paths it observes, and how a set pushed
/// by the server combines with the paths this extension always watches.
///
/// Pure Foundation, with no EndpointSecurity import, so the set arithmetic is unit-testable; FileTamperSubscriber applies the result.
enum WatchedPaths {
    /// builtIn is watched whatever the server pushes. The shipped sudoers rules depend on these paths, so no configuration can take
    /// them away; a pushed set only adds to them.
    static let builtIn: [WatchedPath] = [
        WatchedPath(path: "/etc/sudoers", match: .literal),
        WatchedPath(path: "/etc/sudoers.d/", match: .prefix)
    ]

    /// firmlinkedRoots are the top-level directories macOS keeps under /private and links from the root. Endpoint Security reports
    /// the resolved /private form, and a path is muted in both spellings so an entry works however it was written.
    private static let firmlinkedRoots = ["/etc", "/tmp", "/var"]

    private struct Document: Decodable {
        let version: Int64
        let epoch: Int64?
        let paths: [Entry]
    }

    private struct Entry: Decodable {
        let path: String
        let match: String
    }

    /// decode reads a `watched_paths.update` payload. It returns nil for a payload that is not a watched-path document at all, which
    /// leaves the active set untouched. Within a valid document, an entry the extension would not watch (see isAcceptable) is skipped
    /// and counted rather than failing the whole set, so an older extension keeps watching what it understands when a newer server adds
    /// a kind of entry.
    static func decode(_ data: Data) -> WatchedPathsUpdate? {
        guard let document = try? JSONDecoder().decode(Document.self, from: data) else {
            return nil
        }
        var paths: [WatchedPath] = []
        for entry in document.paths {
            guard let match = WatchedPathMatch(rawValue: entry.match), isAcceptable(entry.path, match) else {
                continue
            }
            paths.append(WatchedPath(path: entry.path, match: match))
        }
        return WatchedPathsUpdate(
            version: document.version, epoch: document.epoch ?? 0, paths: paths, skipped: document.paths.count - paths.count
        )
    }

    /// isAcceptable is the server's rule for one entry (ValidateWatchedPaths in server/rules/api), held again here.
    ///
    /// The server is where a set is validated, but a mute is applied here, and what reaches the kernel is a C string: a path carrying a
    /// NUL, or `..` segments, can read as a deep path to a component count while muting a top-level one, and a top-level prefix puts
    /// every write under that tree on the wire, the firehose ADR-0008 removed after it delayed detection by about 12 minutes. A server
    /// bug should not be able to bring that back, so every rule the server applies to an entry is applied again: an absolute path whose
    /// /private spelling fits within maxPathBytes, free of ASCII control characters (NUL included), with no empty, `.` or `..` segment;
    /// a literal that does not end in "/"; and a prefix that ends in "/" and lies below a top-level directory, judged through /private
    /// for /etc, /tmp and /var.
    static func isAcceptable(_ path: String, _ match: WatchedPathMatch) -> Bool {
        // Every spelling the client mutes must fit, and the /private one of a firmlinked path is the longer.
        guard path.hasPrefix("/"), spellings(of: path).allSatisfy({ $0.utf8.count <= maxPathBytes }),
              !path.unicodeScalars.contains(where: { $0.value < firstPrintable || $0.value == deleteCharacter }) else {
            return false
        }
        let body = path.dropFirst()
        let trimmed = body.hasSuffix("/") ? body.dropLast() : body
        let segments = trimmed.split(separator: "/", omittingEmptySubsequences: false)
        guard !segments.contains(where: { $0.isEmpty || $0 == "." || $0 == ".." }) else {
            return false
        }
        switch match {
        case .literal:
            return !path.hasSuffix("/")
        case .prefix:
            let rootLinked = spellings(of: path).last ?? path
            return path.hasSuffix("/") && rootLinked.split(separator: "/").count >= minimumPrefixDepth
        }
    }

    /// maxPathBytes is the longest path es_mute_path can take: PATH_MAX (1024) counts the terminating NUL of the C string. The server's
    /// MaxWatchedPathBytes.
    private static let maxPathBytes = 1023
    /// firstPrintable and deleteCharacter bound the control characters a path may not carry.
    private static let firstPrintable: UInt32 = 0x20
    private static let deleteCharacter: UInt32 = 0x7f
    /// minimumPrefixDepth is how many path components a prefix needs: a top-level directory is one, so a prefix must name something
    /// inside one.
    private static let minimumPrefixDepth = 2

    /// targets is every path the client mutes for a pushed set: the built-in paths, then the pushed ones, each in both spellings of a
    /// firmlinked root, without duplicates and in a stable order.
    static func targets(pushed: [WatchedPath]) -> [WatchedPath] {
        var seen = Set<WatchedPath>()
        var out: [WatchedPath] = []
        for entry in builtIn + pushed {
            for path in spellings(of: entry.path) {
                let target = WatchedPath(path: path, match: entry.match)
                if seen.insert(target).inserted {
                    out.append(target)
                }
            }
        }
        return out
    }

    /// spellings returns a path in its /private form first, then its root-linked form, when it lies under a firmlinked root, and the
    /// path alone otherwise.
    static func spellings(of path: String) -> [String] {
        for root in firmlinkedRoots {
            if path == root || path.hasPrefix(root + "/") {
                return ["/private" + path, path]
            }
            let privateRoot = "/private" + root
            if path == privateRoot || path.hasPrefix(privateRoot + "/") {
                return [path, String(path.dropFirst("/private".count))]
            }
        }
        return [path]
    }

    /// changes is what moving from the applied targets to the next ones takes: the targets to mute and the targets to unmute, each in
    /// the order of `next` and `applied` respectively. Muting the additions before unmuting the removals means a path in both sets is
    /// never touched, so nothing already watched stops being watched while a set changes.
    static func changes(from applied: [WatchedPath], to next: [WatchedPath]) -> (mute: [WatchedPath], unmute: [WatchedPath]) {
        let appliedSet = Set(applied)
        let nextSet = Set(next)
        return (next.filter { !appliedSet.contains($0) }, applied.filter { !nextSet.contains($0) })
    }
}

/// WatchedPathStore keeps the last accepted set: on disk, so a restarted extension watches it from its first event rather than falling
/// back to the built-in paths until the agent next pushes, and in memory, so a delayed older set is recognised and turned away.
final class WatchedPathStore: Sendable {
    /// defaultStoragePath sits beside the application-control snapshot, in the directory the extension already owns.
    static let defaultStoragePath = "/var/db/com.fleetdm.edr/watched-paths.json"

    let storagePath: String
    private let accepted: OSAllocatedUnfairLock<WatchedPathsUpdate?>

    /// init reads the persisted set, which becomes the baseline a pushed set must supersede.
    init(storagePath: String = WatchedPathStore.defaultStoragePath) {
        self.storagePath = storagePath
        self.accepted = OSAllocatedUnfairLock(initialState: nil)
        let persisted = load()
        accepted.withLock { $0 = persisted }
    }

    /// current is the last accepted set, or nil when none has been.
    var current: WatchedPathsUpdate? {
        accepted.withLock { $0 }
    }

    /// accept decodes a pushed payload and, when it supersedes the last accepted set and has been persisted, records it and returns it
    /// for the caller to apply. It returns nil, changing nothing, for a payload that is not a watched-path document, a set that does not
    /// supersede the current one, or a set that could not be written to disk.
    ///
    /// Persisting comes before accepting so the running set and the one a restart loads never disagree: a set applied without being
    /// written would vanish at the next restart, and, already accepted, could not be retried by a redelivery. Refused instead, it is
    /// retried by the next delivery of the same set.
    ///
    /// Called from the XPC server's serial queue, which is what makes the check, the write and the commit a sequence nothing else
    /// interleaves with.
    func accept(_ data: Data) -> WatchedPathsUpdate? {
        guard let update = WatchedPaths.decode(data), update.supersedes(current), save(data) else {
            return nil
        }
        accepted.withLock { $0 = update }
        return update
    }

    /// load returns the persisted set, or nil when there is none or it cannot be read, in which case the client watches the built-in
    /// paths alone until the agent pushes a set.
    private func load() -> WatchedPathsUpdate? {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: storagePath)) else {
            return nil
        }
        guard let update = WatchedPaths.decode(data) else {
            logger.warning("persisted watched-path set could not be decoded; watching the built-in paths")
            return nil
        }
        return update
    }

    /// save persists a payload exactly as it was pushed, written atomically so a crash mid-write cannot leave a torn file, and reports
    /// whether it was written.
    private func save(_ data: Data) -> Bool {
        do {
            try AtomicFile.write(data, toPath: storagePath)
            return true
        } catch {
            logger.error("watched-path set not applied, persist failed: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }
}

File: server/rules/api/watchedpaths.go
package api

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"
)

// CommandTypeSetWatchedPaths is the command that carries the watched-path set to a host (issue #998). The agent forwards its payload
// to the Endpoint Security extension, whose file-tamper client watches these paths on top of the ones built into it. Stable wire
// string; renaming it breaks every deployed agent.
const CommandTypeSetWatchedPaths = "set_watched_paths"

// WatchedPathMatch is how a watched path covers files. The extension maps each value onto an Endpoint Security target mute type.
type WatchedPathMatch string

const (
	// WatchedPathLiteral covers exactly the file at the path.
	WatchedPathLiteral WatchedPathMatch = "literal"
	// WatchedPathPrefix covers every path that starts with it, at any depth.
	WatchedPathPrefix WatchedPathMatch = "prefix"
)

// WatchedPath is one entry in the watched-path set. The JSON tags are the wire shape the extension decodes (WatchedPaths.swift).
type WatchedPath struct {
	Path  string           `json:"path"`
	Match WatchedPathMatch `json:"match"`
}

// SetWatchedPathsPayload is the payload of a set_watched_paths command. Version is the set's server version, reported back by the
// agent so an operator can see which set a host took. Epoch is the set's update time in Unix microseconds.
//
// The extension applies a set only when its version or its epoch is ahead of the last one it accepted, because commands can reach a
// host out of order. Epoch is what keeps that ordering true after a database restore sends version backwards: the next change is
// stamped with a later wall-clock time than anything before the restore, the same reason application control carries policy_epoch.
type SetWatchedPathsPayload struct {
	Version int64         `json:"version"`
	Epoch   int64         `json:"epoch"`
	Paths   []WatchedPath `json:"paths"`
}

// WatchedPathSet is the stored set, as the REST API reports it.
type WatchedPathSet struct {
	// Version starts at 0, the set no operator has changed, and advances by one on every change.
	Version int64         `json:"version"`
	Paths   []WatchedPath `json:"paths"`
	// UpdatedAt and UpdatedBy are zero until the first change.
	UpdatedAt *time.Time `json:"updated_at,omitempty"`
	UpdatedBy string     `json:"updated_by,omitempty"`
}

// BuiltInWatchedPaths are the paths the extension watches whatever the set holds, mirroring WatchedPaths.builtIn in the extension.
// The set adds to these and cannot remove them. Reported so the console can say what is always watched; the extension, not this
// list, is what enforces it.
var BuiltInWatchedPaths = []WatchedPath{
	{Path: "/etc/sudoers", Match: WatchedPathLiteral},
	{Path: "/etc/sudoers.d/", Match: WatchedPathPrefix},
}

// MaxWatchedPaths bounds the set. Every entry is a mute the file-tamper client holds, two for a path under a firmlinked root, and a
// set this size already covers the uses the issue names (canaries, credential stores, a handful of integrity-monitored directories)
// many times over.
const MaxWatchedPaths = 32

// MaxWatchedPathBytes bounds one path: the macOS PATH_MAX of 1024 counts the terminating NUL of the C string es_mute_path takes, so the
// longest path it can mute is one byte shorter.
const MaxWatchedPathBytes = 1023

// MaxWatchedPathSetBytes bounds the whole set as the server encodes it into a command payload. The fan-out repeats that payload on every
// row of a batched insert of up to 256 hosts, so this keeps one statement near 2 MiB, inside the 4 MiB max_allowed_packet the
// server is designed to work under. Real watched paths are short; the bound only bites on a set of many maximum-length paths, which
// MaxWatchedPaths and MaxWatchedPathBytes alone would allow at over 30 KiB.
const MaxWatchedPathSetBytes = 8 * 1024

// ErrInvalidWatchedPaths is returned for a set the server refuses. The wrapped message names the entry and the reason, and the REST
// handler returns it to the operator.
var ErrInvalidWatchedPaths = errors.New("invalid watched paths")

// ValidateWatchedPaths checks a proposed set. It is the one place the set is validated: the agent checks only the envelope and the
// extension applies what it is given.
//
// A path must be absolute, clean (no empty, "." or ".." segment), within MaxWatchedPathBytes in its /private spelling (the one the
// extension mutes and the kernel reports for /etc, /tmp and /var), and free of ASCII control characters
// (NUL included, which would truncate the path the kernel receives). A prefix
// names a directory, so it ends in "/", and it must lie below a top-level directory: a prefix such as "/Users/" or "/Library/" would
// put every write under that tree on the wire, which is the firehose ADR-0008 removed. A literal names a file, so it does not end in
// "/". An entry may appear once, judged by its root-linked form.
func ValidateWatchedPaths(paths []WatchedPath) error {
	if len(paths) > MaxWatchedPaths {
		return fmt.Errorf("%w: %d paths, at most %d", ErrInvalidWatchedPaths, len(paths), MaxWatchedPaths)
	}
	// Duplicates are keyed by the root-linked form, since /etc/emond.d/ and /private/etc/emond.d/ are one directory.
	seen := make(map[WatchedPath]struct{}, len(paths))
	for i, p := range paths {
		if err := validateWatchedPath(p); err != nil {
			return fmt.Errorf("%w: entry %d (%q): %s", ErrInvalidWatchedPaths, i, p.Path, err.Error())
		}
		key := WatchedPath{Path: rootLinked(p.Path), Match: p.Match}
		if _, dup := seen[key]; dup {
			return fmt.Errorf("%w: entry %d (%q): listed more than once", ErrInvalidWatchedPaths, i, p.Path)
		}
		seen[key] = struct{}{}
	}
	// Measured with the encoder the payload is written with, so JSON escaping (which can grow a byte to six) is counted as sent.
	encoded, _ := json.Marshal(paths)
	if len(encoded) > MaxWatchedPathSetBytes {
		return fmt.Errorf("%w: the set encodes to %d bytes, at most %d", ErrInvalidWatchedPaths, len(encoded), MaxWatchedPathSetBytes)
	}
	return nil
}

func validateWatchedPath(p WatchedPath) error {
	switch {
	case !strings.HasPrefix(p.Path, "/"):
		return errors.New("the path must be absolute")
	case len(privateSpelling(p.Path)) > MaxWatchedPathBytes:
		// The extension also mutes a firmlinked path in its /private spelling, the one the kernel reports, which is the longer.
		return fmt.Errorf("the path is longer than %d bytes in its /private spelling", MaxWatchedPathBytes)
	case strings.ContainsFunc(p.Path, func(r rune) bool { return r < 0x20 || r == 0x7f }):
		return errors.New("the path contains an ASCII control character")
	}
	for s := range strings.SplitSeq(strings.TrimSuffix(p.Path[1:], "/"), "/") {
		if s == "" || s == "." || s == ".." {
			return errors.New(`the path has an empty, "." or ".." segment`)
		}
	}
	switch p.Match {
	case WatchedPathLiteral:
		if strings.HasSuffix(p.Path, "/") {
			return errors.New(`a literal names a file, so it must not end in "/"`)
		}
	case WatchedPathPrefix:
		if !strings.HasSuffix(p.Path, "/") {
			return errors.New(`a prefix names a directory, so it must end in "/"`)
		}
		if len(strings.Split(strings.TrimSuffix(rootLinked(p.Path)[1:], "/"), "/")) < 2 {
			return errors.New("a prefix must lie below a top-level directory")
		}
	default:
		return fmt.Errorf("unknown match %q, want %q or %q", p.Match, WatchedPathLiteral, WatchedPathPrefix)
	}
	return nil
}

// privateSpelling returns a path under /etc, /tmp or /var in its /private form, and any other path unchanged. The mapping itself is
// macOSPathAlias's, which detection-config path matching also uses, so the two cannot disagree about which paths are firmlinked.
func privateSpelling(path string) string {
	if alias := macOSPathAlias(path); strings.HasPrefix(alias, "/private/") {
		return alias
	}
	return path
}

// rootLinked returns a path under /private/etc, /private/tmp or /private/var in its root-linked form, and any other path unchanged.
func rootLinked(path string) string {
	if alias := macOSPathAlias(path); alias != "" && !strings.HasPrefix(alias, "/private/") {
		return alias
	}
	return path
}
